In [1]:
## Standard Imports
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
## Importing the OS and JSON Modules
import os,json

In [2]:
info_df = pd.read_csv('Data/superhero_info - superhero_info.csv')
info_df.head()

Hero|Publisher Gender               Race Alignment  \
0          A-Bomb|Marvel Comics   Male              Human      good   
1  Abe Sapien|Dark Horse Comics   Male      Icthyo Sapien      good   
2            Abin Sur|DC Comics   Male            Ungaran      good   
3     Abomination|Marvel Comics   Male  Human / Radiation       bad   
4   Absorbing Man|Marvel Comics   Male              Human       bad   

  Hair color Eye color Skin color  \
0    No Hair    yellow    Unknown   
1    No Hair      blue       blue   
2    No Hair      blue        red   
3    No Hair     green    Unknown   
4    No Hair      blue    Unknown   

                                   Measurements  
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}  
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}  
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}  
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}  
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}

In [3]:
info_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 463 entries, 0 to 462
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Hero|Publisher  463 non-null    object
 1   Gender          463 non-null    object
 2   Race            463 non-null    object
 3   Alignment       463 non-null    object
 4   Hair color      463 non-null    object
 5   Eye color       463 non-null    object
 6   Skin color      463 non-null    object
 7   Measurements    463 non-null    object
dtypes: object(8)
memory usage: 29.1+ KB


In [4]:
sup_pow_df = pd.read_csv('Data/superhero_powers - superhero_powers.csv')
sup_pow_df.head()

hero_names                                             Powers
0      3-D Man         Agility,Super Strength,Stamina,Super Speed
1       A-Bomb  Accelerated Healing,Durability,Longevity,Super...
2   Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...
3     Abin Sur                                 Lantern Power Ring
4  Abomination  Accelerated Healing,Intelligence,Super Strengt...

In [5]:
sup_pow_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 667 entries, 0 to 666
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   hero_names  667 non-null    object
 1   Powers      667 non-null    object
dtypes: object(2)
memory usage: 10.5+ KB


In [6]:
#name and publisher
info_df['Hero|Publisher'].head()

0            A-Bomb|Marvel Comics
1    Abe Sapien|Dark Horse Comics
2              Abin Sur|DC Comics
3       Abomination|Marvel Comics
4     Absorbing Man|Marvel Comics
Name: Hero|Publisher, dtype: object

In [7]:
## adding expand=True
info_df[['Hero Name', 'Publisher']] = info_df\
    ['Hero|Publisher'].str.split('|',expand=True)
info_df = info_df.drop(columns = ['Hero|Publisher'])
info_df.head(3)

Gender           Race Alignment Hair color Eye color Skin color  \
0   Male          Human      good    No Hair    yellow    Unknown   
1   Male  Icthyo Sapien      good    No Hair      blue       blue   
2   Male        Ungaran      good    No Hair      blue        red   

                                   Measurements   Hero Name          Publisher  
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}      A-Bomb      Marvel Comics  
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}  Abe Sapien  Dark Horse Comics  
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}    Abin Sur          DC Comics

In [8]:
print(type(info_df.loc[0, 'Measurements']))
info_df.loc[0, 'Measurements']

<class 'str'>


"{'Height': '203.0 cm', 'Weight': '441.0 kg'}"

In [9]:
info_df['Measurements'] = info_df['Measurements'].str\
            .replace("'", '"')
info_df['Measurements'] = info_df['Measurements']\
            .apply(json.loads)
info_df.head(3)

Gender           Race Alignment Hair color Eye color Skin color  \
0   Male          Human      good    No Hair    yellow    Unknown   
1   Male  Icthyo Sapien      good    No Hair      blue       blue   
2   Male        Ungaran      good    No Hair      blue        red   

                                   Measurements   Hero Name          Publisher  
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}      A-Bomb      Marvel Comics  
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}  Abe Sapien  Dark Horse Comics  
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}    Abin Sur          DC Comics

In [10]:
print(type(info_df.loc[0, 'Measurements']))
info_df.loc[0, 'Measurements']

<class 'dict'>


{'Height': '203.0 cm', 'Weight': '441.0 kg'}

In [11]:
measurements = info_df['Measurements'].apply(pd.Series)
measurements

Height    Weight
0    203.0 cm  441.0 kg
1    191.0 cm   65.0 kg
2    185.0 cm   90.0 kg
3    203.0 cm  441.0 kg
4    193.0 cm  122.0 kg
..        ...       ...
458  183.0 cm   83.0 kg
459  165.0 cm   52.0 kg
460   66.0 cm   17.0 kg
461  170.0 cm   57.0 kg
462  185.0 cm   81.0 kg

[463 rows x 2 columns]

In [12]:
info_df = pd.concat((info_df, measurements), axis = 1)
info_df = info_df.drop(columns = ['Measurements'])

In [13]:
to_replace = [' cm',' kg']

for char in to_replace:
    info_df['Height'] = info_df['Height']\
                    .str.replace(char, '', regex=False)
    info_df['Weight'] = info_df['Weight']\
                    .str.replace(char, '', regex=False)

columns = ['Height', 'Weight']
for col in columns:
    info_df[col] = pd.to_numeric(info_df[col])

In [14]:
info_df = info_df.reindex(columns=[\
            'Hero Name', 'Publisher', 'Gender', 'Eye color',\
            'Race', 'Hair color', 'Height', 'Skin color',\
            'Alignment', 'Weight'])
info_df.head(3)

Hero Name          Publisher Gender Eye color           Race Hair color  \
0      A-Bomb      Marvel Comics   Male    yellow          Human    No Hair   
1  Abe Sapien  Dark Horse Comics   Male      blue  Icthyo Sapien    No Hair   
2    Abin Sur          DC Comics   Male      blue        Ungaran    No Hair   

   Height Skin color Alignment  Weight  
0   203.0    Unknown      good   441.0  
1   191.0       blue      good    65.0  
2   185.0        red      good    90.0

In [15]:
#Combining DF
hero_pow = sup_pow_df['hero_names']\
            .isin(info_df['Hero Name'])

sup_pow_df = sup_pow_df[hero_pow]
sup_pow_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 455 entries, 1 to 666
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   hero_names  455 non-null    object
 1   Powers      455 non-null    object
dtypes: object(2)
memory usage: 10.7+ KB


In [16]:
sup_pow_df = sup_pow_df.rename(columns =\
                                    {'hero_names':'Hero Name'})

In [17]:
merg_sup_df = pd.merge(info_df, sup_pow_df,\
                          on='Hero Name', how='inner')
merg_sup_df.head(3)

Hero Name          Publisher Gender Eye color           Race Hair color  \
0      A-Bomb      Marvel Comics   Male    yellow          Human    No Hair   
1  Abe Sapien  Dark Horse Comics   Male      blue  Icthyo Sapien    No Hair   
2    Abin Sur          DC Comics   Male      blue        Ungaran    No Hair   

   Height Skin color Alignment  Weight  \
0   203.0    Unknown      good   441.0   
1   191.0       blue      good    65.0   
2   185.0        red      good    90.0   

                                              Powers  
0  Accelerated Healing,Durability,Longevity,Super...  
1  Agility,Accelerated Healing,Cold Resistance,Du...  
2                                 Lantern Power Ring

In [18]:
#one-hot-encoded columns for every power that appears in the dataset
print(type(merg_sup_df.loc[0,'Powers']))
merg_sup_df.loc[0,'Powers']

<class 'str'>


'Accelerated Healing,Durability,Longevity,Super Strength,Stamina,Camouflage,Self-Sustenance'

In [19]:
spl_pow_df = merg_sup_df['Powers'].str.split(",")

In [20]:
exp= spl_pow_df.explode('Powers')
exp.head(3)

0    Accelerated Healing
1             Durability
2              Longevity
Name: Powers, dtype: object

In [21]:
#saving unique values from exploded column
colum = exp.dropna().unique()
colum

array(['Accelerated Healing', 'Durability', 'Longevity', 'Super Strength',
       'Stamina', 'Camouflage', 'Self-Sustenance', 'Agility',
       'Cold Resistance', 'Underwater breathing', 'Marksmanship',
       'Weapons Master', 'Intelligence', 'Telepathy', 'Immortality',
       'Reflexes', 'Enhanced Sight', 'Sub-Mariner', 'Lantern Power Ring',
       'Super Speed', 'Invulnerability', 'Animation', 'Super Breath',
       'Energy Absorption', 'Elemental Transmogrification',
       'Fire Resistance', 'Natural Armor', 'Molecular Manipulation',
       'Heat Resistance', 'Matter Absorption', 'Stealth', 'Flight',
       'Power Suit', 'Energy Blasts', 'Energy Beams', 'Power Cosmic',
       'Heat Generation', 'Danger Sense', 'Teleportation', 'Phasing',
       'Force Fields', 'Hypnokinesis', 'Energy Manipulation',
       'Invisibility', 'Enhanced Senses', 'Jump', 'Substance Secretion',
       'Natural Weapons', 'Wallcrawling', 'Vision - Thermal',
       'Power Augmentation', 'Cryokinesis', 'Dupli

In [22]:
for col in colum:
    merg_sup_df[col] = merg_sup_df['Powers'].str.contains(col)
merg_sup_df.head()

/var/folders/bt/3cwrk5g14pn7211jcqqqp7xw0000gn/T/ipykernel_22141/438110512.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  merg_sup_df[col] = merg_sup_df['Powers'].str.contains(col)
/var/folders/bt/3cwrk5g14pn7211jcqqqp7xw0000gn/T/ipykernel_22141/438110512.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  merg_sup_df[col] = merg_sup_df['Powers'].str.contains(col)
/var/folders/bt/3cwrk5g14pn7211jcqqqp7xw0000gn/T/ipykernel_22141/438110512.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result o

/var/folders/bt/3cwrk5g14pn7211jcqqqp7xw0000gn/T/ipykernel_22141/438110512.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  merg_sup_df[col] = merg_sup_df['Powers'].str.contains(col)
/var/folders/bt/3cwrk5g14pn7211jcqqqp7xw0000gn/T/ipykernel_22141/438110512.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  merg_sup_df[col] = merg_sup_df['Powers'].str.contains(col)
/var/folders/bt/3cwrk5g14pn7211jcqqqp7xw0000gn/T/ipykernel_22141/438110512.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result o

/var/folders/bt/3cwrk5g14pn7211jcqqqp7xw0000gn/T/ipykernel_22141/438110512.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  merg_sup_df[col] = merg_sup_df['Powers'].str.contains(col)
/var/folders/bt/3cwrk5g14pn7211jcqqqp7xw0000gn/T/ipykernel_22141/438110512.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  merg_sup_df[col] = merg_sup_df['Powers'].str.contains(col)
/var/folders/bt/3cwrk5g14pn7211jcqqqp7xw0000gn/T/ipykernel_22141/438110512.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result o

Hero Name          Publisher Gender Eye color               Race  \
0         A-Bomb      Marvel Comics   Male    yellow              Human   
1     Abe Sapien  Dark Horse Comics   Male      blue      Icthyo Sapien   
2       Abin Sur          DC Comics   Male      blue            Ungaran   
3    Abomination      Marvel Comics   Male     green  Human / Radiation   
4  Absorbing Man      Marvel Comics   Male      blue              Human   

  Hair color  Height Skin color Alignment  Weight  ... Hair Manipulation  \
0    No Hair   203.0    Unknown      good   441.0  ...             False   
1    No Hair   191.0       blue      good    65.0  ...             False   
2    No Hair   185.0        red      good    90.0  ...             False   
3    No Hair   203.0    Unknown       bad   441.0  ...             False   
4    No Hair   193.0    Unknown       bad   122.0  ...             False   

   Weather Control  Nova Force  Odin Force  Phoenix Force  Power Sense  \
0            False       False       False          False        False   
1            False       False       False          False        False   
2            False       False       False          False        False   
3            False       False       False          False        False   
4            False       False       False          False        False   

   Qwardian Power Ring  Melting  Changing Armor  Terrakinesis  
0                False    False           False         False  
1                False    False           False         False  
2                False    False           False         False  
3                False    False           False         False  
4                False    False           False         False  

[5 rows x 170 columns]

## Questions


1. Compare the average weight of super powers who have Super Speed to those who do not.

In [27]:
merg_sup_df.groupby('Super Speed')['Weight'].mean()

Super Speed
False    101.773585
True     129.404040
Name: Weight, dtype: float64

- The average weight of those that HAVE super speed would be 129.40 kg 
- The average weight of those that do NOT have super speed would be 101.77 kg

2. What is the average height of heroes for each publisher?

In [28]:
merg_sup_df.groupby('Publisher')['Height'].mean()

Publisher
DC Comics            181.923913
Dark Horse Comics    176.909091
George Lucas         159.600000
Image Comics         211.000000
Marvel Comics        191.546128
Shueisha             171.500000
Star Trek            181.500000
Team Epic TV         180.750000
Unknown              178.000000
Name: Height, dtype: float64

- Above, is the average height of heroes for each publisher. 